### PR TITLE
Inbound ratelimit middleware

### PR DESCRIPTION
**Pull request by kriskowal**
This change introduces an inbound rate limiter middleware. This applies a rate limit closing over all inbound requests. The intention is for this feature to become public API. The next change in this stack will introduce a config shape.

### CODE REVIEW

File: internal/ratelimit/ratelimit.go [hunk 1]
```diff
@@ -73,13 +74,21 @@ func NewThrottle(rps int, opts ...Option) *Throttle {
 		options.clock = clock.NewReal()
 	}
 
+	if rps <= 0 {
```
Comment (willhug) [inline internal/ratelimit/ratelimit.go]: can we just make this a uint?

Comment (kriskowal) [inline internal/ratelimit/ratelimit.go]: I think I will leave it as is. Internally it’s casting into an int64, but I already made the choice to use a plain int at the interface for simplicity. Folks sometimes do math with these numbers, which is just easier to do when all terms are ints.

**Comment by kriskowal** [comment]
Yeah, I expect that there will be policies in the future, which we can add to this config. There will still be a blanket rate limit for the whole task.

File: internal/integrationtest/util.go [hunk 2]
```diff
@@ -211,6 +207,9 @@ func CallUntilSuccess(t *testing.T, rawClient raw.Client, interval time.Duration
 
 // Call sends an echo request to the client.
 func Call(ctx context.Context, rawClient raw.Client) error {
+	if ctx == nil {
```
Comment (peter-edge) [inline internal/integrationtest/util.go]: This is a little idk, contexts are generally expected to not be nil right?

File: internal/integrationtest/util.go [hunk 3]
```diff
@@ -222,3 +221,11 @@ func Call(ctx context.Context, rawClient raw.Client) error {
 	}
 	return nil
 }
+
+// Register registers an echo procedure handler on a dispatcher.
+func Register(dispatcher *yarpc.Dispatcher) {
+	handle := func(ctx context.Context, req []byte) ([]byte, error) {
```
Comment (peter-edge) [inline internal/integrationtest/util.go]: Could just inline this :-)

File: internal/ratelimit/inbound_middleware.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	throttle *Throttle
+}
+
+var _ middleware.UnaryInbound = (*UnaryInboundMiddleware)(nil)
+
+// Handle drops inbound requests with a ResourceExhaustedError if the arrive
+// more frequently than the configured rate limit.
+func (m *UnaryInboundMiddleware) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, next transport.UnaryHandler) error {
+	if m.throttle.Throttle() {
+		return yarpcerrors.ResourceExhaustedErrorf("rate limit exceeded")
```
Comment (peter-edge) [inline internal/ratelimit/inbound_middleware.go]: nit: could make this a constant :-)

File: internal/integrationtest/util.go (hunk 2 above)
Comment (kriskowal) [inline internal/integrationtest/util.go]: Yeh, and linter complained so I had to pass a background. This is not exercised anymore.

File: internal/ratelimit/inbound_middleware.go (hunk 4 above)
Comment (kriskowal) [inline internal/ratelimit/inbound_middleware.go]: Yeah, this is actually important.

**Pull request merged**